wire-ios-request-strategy/Sources/Request Strategies/MLS/Actions/UpdateConversationProtocolActionHandler.swift

```diff
@@ -0,0 +1,72 @@
+//
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import WireDataModel
+
+final class UpdateConversationProtocolActionHandler: ActionHandler<UpdateConversationProtocolAction> {
+
+    typealias EventPayload = [AnyHashable: Any]
+
+    // MARK: - Methods
+
+    override func request(
+        for action: Action,
+        apiVersion: APIVersion
+    ) -> ZMTransportRequest? {
+
+        var action = action
+
+        guard apiVersion >= .v5 else {
+            action.fail(with: .endpointUnavailable)
+            return .none
+        }
+
+        let domain = action.qualifiedID.domain
+        let conversationID = action.qualifiedID.uuid.transportString()
+        let messageProtocol = action.messageProtocol.stringValue
+        let path = "/conversations/\(domain)/\(conversationID)/\(messageProtocol)"
+        let payload = ["protocol": action.messageProtocol.stringValue] as ZMTransportData
+
+        return .init(
+            path: path,
+            method: .put,
+            payload: payload,
+            apiVersion: apiVersion.rawValue
+        )
+    }
+
+    override func handleResponse(
+        _ response: ZMTransportResponse,
+        action: Action
+    ) {
+        var action = action
+
+        if [200, 204].contains(response.httpStatus) {
+            return action.succeed()
+        }
+
+        if
+            let label = response.payloadLabel(),
+            let apiFailure = Action.Failure.APIFailure(rawValue: label),
+            apiFailure.statusCode == response.httpStatus {
```

Comment:
This only works under the condition that there's a label in the payload. Unfortunately this is not always the case, there could be a status but no label. e.g: status `404` without a label means the conversation or the domain wasn't found. (sadly this is not obvious in swagger)

I'd suggest to do something like this 

```swift
switch (response.httpStatus, response.payloadLabel()) {
   case (200, _), (204, _):
      action.succeed()
   case (400, "mis-migration-criteria-not-satisfied"):
      action.fail(with: .api(.mlsMigrationCriteriaNotSatisfied))
   case (400, _):
      action.fail(with: .api(.invalidBody))
   case (404, "no-team"):
      action.fail(with: .api(.noTeam))
   case (404, _):
      action.fail(with: .api(.conversationIDOrDomainNotFound))
   // ...
   default:
      action.fail(with: .unknown)
}
```

Doing the mapping like this in the handler also ensure that responsibilities are kept separate between the action and the handler. The action only concerns itself with the expected input and output, and the handler can take care of creating the API request and mapping the response

Reply:
> > This only works under the condition that there's a label in the payload. Unfortunately this is not always the case
> 
> I see.
> 
> @typfel Then the code which you mentioned is probably also not correct, since it doesn't consider the status code at all, right?
> 
> https://github.com/wireapp/wire-ios/blob/eff95a8097bd8f1ada0e1443f1a4d3dccca5d1df/wire-ios-request-strategy/Sources/Request%20Strategies/Connection/Actions/UpdateConnectionActionHandler.swift#L100-L102
> 
> 
> Maybe in other places as well.

No, it doesn't but the status code is also encoded inside the `Payload.ResponseFailure`, but never mind let go with what David suggested since that's the system in place now for action handlers.